### ISSUE TITLE
MySQL5.7 support

### ISSUE THREAD

**Issue by jekkos** [jekkos/opensourcepos#630]
A bug I spotted this morning, seems to be an issue with the Item->search query that was refactored not so long ago.
1. Execute the item import
2. Table is not refreshed (I saw a status 500 coming from item search)
3. Seems to be a problem with a query select in combination with group by (and company name column)

**Comment by daN4cat** [comment]
What do you mean with "Item import"?

**Comment by jekkos** [comment]
Item excel import. Could be a db issue as well (as I'm testing on a containerized ospos now)

```ERROR 1055 (42000): Expression #26 of SELECT list is not in GROUP BY clause and contains nonaggregated column 'ospos.ospos_suppliers.company_name' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by```

**Comment by daN4cat** [comment]
I fixed the excel imports issues due to the lack of notifications and also not translated strings, therefore I tested different cases with Customers and Items and I couldn't get anything unusual.

I'll check better later and see if something springs to my mind around that issue.

**Comment by jekkos** [comment]
Could be mysql config issue as the item search works fine with my  regular install. Maybe something new in Mysql 5.7, will need to check that (also for #616)

**Comment by daN4cat** [comment]
Yes, it's a MySQL 5.7 (new) thing: http://stackoverflow.com/questions/36228836/syntax-error-or-access-violation-1055-expression-8-of-select-list-is-not-in-gr

**Comment by daN4cat** [comment]
Here an upgrade to MySQL5.7 guideline: https://www.digitalocean.com/community/tutorials/how-to-prepare-for-your-mysql-5-7-upgrade

**Comment by daN4cat** [comment]
I moved this to 3.1 because MySQL5.7 support is something for later, like PHP7.

**Comment by daN4cat** [comment]
Although one could argue that the issue is not a proper behaviour anyway, besides MySQL5.7.

**Comment by jekkos** [comment]
In other db's you need to add select field to the group by as well if you want to have it in the resultset. Mysql seemed to be an exception to this (which is no longer true with 5.7)

**Comment by daN4cat** [comment]
Yes, this requires a review of all the model and SQL queries.
Quite a job, but later now I want to get to an end with XSS clean (almost done with Sales and I have only Receiving left after that).

**Comment by SteveIreland** [comment]
This falls comfortably within my skill set, so I'd like to help out on this issue - if you're accepting help,

**Comment by odiea** [comment]
That would be great.

**Comment by daN4cat** [comment]
The problem is in the models and probably the fact that table joints are not explicitly mentioning the table name.field. That's what I understood at the time but had no time to work on this.
So look into models folder as a starting point.
I know I need to sanity check temp table and in specific receiving one.

**Comment by daN4cat** [comment]
And thanks for the help.

**Comment by SteveIreland** [comment]
I'm glad to help and hope to be around for awhile.  I've already committed to install, configure and support OSPOS for a friend for his small business operations. So ..."I'm all in!".

**Comment by daN4cat** [comment]
Sounds great, we need help.
So this is a good starting point to learn a bit the code.

**Comment by daN4cat** [comment]
Please bear in mind that although we made major improvements there is legacy code that might not be the best, so if you feel like something should be better let's discuss.

**Comment by SteveIreland** [comment]
Right.  I'm not planning on rewriting anything.  The first step is to adapt the MySQL 5.6 compliant "group by" clauses to be 5.7 compliant.  To me it looks like I'm only facing, as a worst case scenario, about 17 PHP file changes in order to eliminate the group by issue - which isn't bad at all.  And since I don't want to affect the behavior of the application I'll just make the query "less ambiguous" even if the code might actually need to be reworked a bit.  If I see anything that troubles my soul I'll stop and ask for direction before proceeding.

**Comment by daN4cat** [comment]
Sounds good, thanks.

**Comment by daN4cat** [comment]
So we can probably make this change for 3.0.1

**Comment by jekkos** [comment]
OK would be good to have it fixed in the code indeed. A quick solution would be disabling ONLY_FULL_GROUP_BY option in mysql but that requires adapting the config file. Think it's better to write the queries correctly so that there are no results 'made up' by Mysql

**Comment by jekkos** [comment]
Also see #938

**Comment by jekkos** [comment]
@daN4cat let's keep 3.0.1 for minor things + improvements and release a bigger new feat in 3.1 (like categories)?

**Comment by daN4cat** [comment]
I was considering MySQL5.7 a small thing and improvement, let see how it goes and decide later. I left 3.1.0 as target version.

**Comment by daN4cat** [comment]
@SteveIreland how is going? Are you close to completion?
I'm still wondering if this could make 3.0.1 if not a big change so we avoid issues reporting as we had in recent times...

**Comment by SteveIreland** [comment]
When I committed I was pretty confident that it would be a simple correction to an armful of SQL statements. I competed the code changes I was expecting to make,  but I stumbled when I started testing.  On the Item panel, the SQL statement that CodeIgniter generates works when I run it in MySQL  Workbench, but no items are being listed.  So I've been deep diving into OSPOS and CodeIgniter to determine where it's breaking down. Before I committed I saw the ticket was already 5 months old and I checked the milestone date for 3.0.1 and saw that it was due December 31st so I assumed I had a bit of time to get up to speed.  However, if this needs to move more quickly then please don't wait on me - I will NOT be offended.  Based on what I've been swimming through I definitely have a bit more to learn about how the product is wired together.

**Comment by daN4cat** [comment]
Thanks for the work and indeed it's not always trivial to change database interfacing.
I don't know if you are aware but now with 3.0.1 there is a means to dump the query logs in application/log/. Just open config/config.php and look for the configured item related to that and set it TRUE.

No worries, it's still tagged 3.1.0 so we leave it as it is and wait.

I need to fix 3.0.1 deadline and set end of this month.

**Comment by SteveIreland** [comment]
I finally made it past the hiccup I had.  It turns out that on a multi-file group by join since there is a default"*" for the list of fields to be retrieved I really needed to "instantiate/identify" each of the fields  from ALL of the tables (I was hoping that I could just get away with the fields from the primary table) - otherwise something upstream might be broken (which it is was in this particular case).  As Francesco points out the fix is trivial.   So what's my next step?  Unfortunately I don't yet have a live site to use to test my changes and I don't want to break any one else's site.  How should I approach publishing my changes?   Is there a QA process in place that will identify where I made any mistakes?

**Comment by odiea** [comment]
Have you tried wamp,xampp or a vm?

**Comment by SteveIreland** [comment]
Sorry, I'm using WAMP and have swap between MySQL 5.7 and MySQL 5.6 for testing to make sure my changes work the same for both database versions.  I load up the files with small sets of records but being a newbie here I just wanted to make sure sure that I was aware of any formal test scripts that contributors are required to run through.  If not then I'll start setting up my own test scenarios and start exercising the changes.  Thanks!

**Comment by daN4cat** [comment]
@SteveIreland you should go for a PR (pull request) and then we can review and test and if fine we integrate.

I hope you worked on the latest 3.0.1 because I optimised the database for the reports.

**Comment by jekkos** [comment]
@SteveIreland no test scenario's yet for these models, you might need to do some sanity check first and then let community test what's left after the merge.

**Comment by odiea** [comment]
I created a Ubuntu server with mysql 5.7. So far the main issue is the tables for Takings and Items do not load. Create Item works just fine. Is this the issue that is going on with this mysql version. @SteveIreland  if you would like to submit a zip of the changes you think will fix it let me know.

**Comment by SteveIreland** [comment]
I've been merging the changes in from the jekkos:master branch before I do any of my own changes or testing so  that includes 3.01 ... right?  I don't see a specific 3.0.1 branch.  I see potential for extreme embarrassment on my part otherwise.

**Comment by odiea** [comment]
I am willing to pretest this. That is what I kinda do here. If it does not work I will let you know. Do not be embarrassed the best have made mistakes.

**Comment by odiea** [comment]
I am also getting the error that the image file upload folder may not be writeable. They are set at 0755 same as my other Linux servers. Possibly another 5.7 issue?

**Comment by SteveIreland** [comment]
Thanks @odiea.  I'm still working through some of the quirky behavior of CodeIgniter's query builder.  It seems to want to do some odd things when the SQL is "out of the ordinary".  I haven't explored the image side yet so I can't comment on your image issue, but I haven't yet stumbled across any documentation or reports suggesting that storing images in the MySQL database has changed in 5.7, but hopefully OSPOS doesn't store images in the database.

**Comment by daN4cat** [comment]
Images (= pictures) are not stored in MySQL for OSPOS case.

**Comment by SteveIreland** [comment]
I submitted a pull request for this.  This includes a change to all of the php files involving group_by functions.  Only Item.php and Sales.php were significantly changed.  There were several others changed but they were trivial.

**Comment by daN4cat** [comment]
As discussed in #989 we use in 3.0.1 `SET GLOBAL sql_mode=(SELECT REPLACE(@@sql_mode,'ONLY_FULL_GROUP_BY',''));` to allow people to use MySQL5.7 with existing database queries.

We still aim at support MySQL5.7 properly in 3.1 as per @SteveIreland work.

**Comment by odiea** [comment]
Yes. I do understand what is going on. It was nice to see that the mysql5.7 was causing the image upload issues and not folder permissions.

**Comment by daN4cat** [comment]
As soon as I get confirmation that 3.0.1 temp fix is fine I will look into @SteveIreland changes and merge them in to start working on the 3.1.
There are 3 PR pending now but they are all for 3.1 and there will be a bit of integration work for two of them.

**Comment by jekkos** [comment]
Yes as @odiea reported the sql mode might not be persisted across server restart. We might want to set it in the code somewhere.

**Comment by daN4cat** [comment]
I'm afraid I missed that, therefore I need to revert the database script and see if we can do something else from code. Otherwise I will remove the MySQL5.7 support and just release 3.0.1. I don't want to hold that back anymore.

**Comment by jekkos** [comment]
Maybe we can add some kind of post-connect hook that does this for us. Let me check the docs.

**Comment by jekkos** [comment]
https://github.com/bcit-ci/CodeIgniter/pull/4349

**Comment by odiea** [comment]
'options' => array(
        MYSQLI_INIT_COMMAND => "SET SESSION sql_mode = 'NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION'"
    )
Added this to the same location and it did not help.

**Comment by odiea** [comment]
@jekkos mysqli_driver.php. Can any of this be altered to achieve what is needed?
if (isset($this->stricton))
		{
			if ($this->stricton)
			{
				$this->_mysqli->options(MYSQLI_INIT_COMMAND, 'SET SESSION sql_mode = CONCAT(@@sql_mode, ",", "STRICT_ALL_TABLES")');
			}
			else
			{
				$this->_mysqli->options(MYSQLI_INIT_COMMAND,
					'SET SESSION sql_mode =
					REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(
					@@sql_mode,
					"STRICT_ALL_TABLES,", ""),
					",STRICT_ALL_TABLES", ""),
					"STRICT_ALL_TABLES", ""),
					"STRICT_TRANS_TABLES,", ""),
					",STRICT_TRANS_TABLES", ""),
					"STRICT_TRANS_TABLES", "")'
				);
			}
		}

**Comment by odiea** [comment]
What happens if you change stricton to true in the database config.

**Comment by odiea** [comment]
I will give this a try on the 5.7 ubuntu server and let you know what happened. Later today.

**Comment by jekkos** [comment]
Hm I checked that code too indeed. I'm not sure whether we can fix this by using CI's builtin driver configuration. There might be a way if we customize the driver ourselves as written in the stack overflow post.

another option would have been the post connect hook but as most connections are pooled I doesn't seem to be a good idea? 

So maybe we should drop the idea of backward compatibility as far as I can see it can only easily be configured by adapting mysql configuration itself.

**Comment by daN4cat** [comment]
I revert that change and we fix properly in 3.1.0

**Comment by jekkos** [comment]
Ok thanks .

**Comment by daN4cat** [comment]
Done

**Comment by odiea** [comment]
mysqli_driver.php. I changed the line 143 to the following. Then I set stricton to True. Now tables show with no errors. Change back to false when work is completed.

$this->_mysqli->options(MYSQLI_INIT_COMMAND, 'SET SESSION sql_mode = "STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION"');

**Comment by odiea** [comment]
Leaving 'stricton' => TRUE, with the current mysql lower than 5.7 does not appear to have any effects on the current setup. Changing this back to false with 5.7 now shows errors and tables do not load.

**Comment by SteveIreland** [comment]
So .... after all is said and done, should I correct the errors and resubmit?  daN4cat found one mistake and I'm sure there will be more.  I should have spotted that - sorry.  Also, I didn't touch the raw SQL "group by" statement's yet.  In my first draft I did, but there were so many repeated conflicts I was scared off and decided to just focus on the group_by functions.  So in the next iteration I'll address both ...  just need a bit of guidance from the deciders.  :-)

**Comment by daN4cat** [comment]
@SteveIreland yes, a bit of a debate and no proper attention to the issue from my side too.
I looked better into the MySQL5.7 changes reading this: https://www.digitalocean.com/community/tutorials/how-to-prepare-for-your-mysql-5-7-upgrade

And besides the GROUP BY in point 5, there is more to check. For instance point 3 about the string size, I'm not sure there are proper guarding controls in the code to avoid that. I bet that if you enter somewhere a comment longer than the field size we get an error.

So we should really check each point from the link and make sure the code is fine.

**Comment by jekkos** [comment]
The group by story is something we should look into. As reported in #938 Mysql would make up values in some situations in the past. It does not make sense to use group by without aggregates, so we should try to clean up those cases. I knew this would require some thorough review so thanks for the link @daN4cat

**Comment by odiea** [comment]
So no one going to try this temp fix.

**Comment by daN4cat** [comment]
@odiea not sure that's the only fix we need and it's better to do a proper job rather than rush and issue an half baked one

**Comment by daN4cat** [comment]
@SteveIreland I noticed you closed the PR and deleted the branch. Anything to say about that?
The work you have done was a good starting point and we just need to elaborate more.
I hope your intention is to have a new branch with your original code and more rework.

**Comment by jekkos** [comment]
@SteveIreland Don't leave us! I wanted to at least fetch what you did and continue the effort to solve this issue.

**Comment by daN4cat** [comment]
@jekkos I have it in a local branch at least. I will commit it in master as mysql57 branch

**Comment by SteveIreland** [comment]
Yes, I'm rebuilding the branch.  Sorry for the slow response.  We are experiencing an annual time and energy sucking festival called Thanksgiving here in the States - since I'm the patriarch all demands are my responsibility to satisfy.  Anyway, I hope to publish a slightly improved branch by the end of this weekend.

**Comment by jekkos** [comment]
Aha! The holy day of consumerism. Don't feel sorry, take your time we still have some work first to round up current 3.0.1 iteration.

**Comment by daN4cat** [comment]
Got caught in the Black Friday online bonanza spending the outrageous amount of £20 for a 128GB SD card on amazon.
My wife with little discount on just selected items had a crazy busy day yesterday in her shop.
UK is becoming the 51st state.... what do we do with the Queen?

**Comment by daN4cat** [comment]
@SteveIreland please rebase your PR with the latest master because Release 3.0.1 is now out.
I created a 3.0.2 release label but I will not touch master unless something is really required to issue a 3.0.2.
I prefer to go for a 3.1 with all the new PR in it.

**Comment by SteveIreland** [comment]
Thanks, I'm working on that now.  I was having a problem getting through the tests on the sales reports and decided that I must not have the latest and greatest version installed since I didn't seen any error reports.  So I resync'd just a few minutes ago and the sales reports are working again.  So I'll re-merge again and reapply my changes and retest again to see how close I'm getting.

**Comment by SteveIreland** [comment]
I believe I've submitted another pull request.  Please let me know if it wasn't successful or if what I've done is totally useless.  Thanks.

**Comment by daN4cat** [comment]
Are we confident MySQL5.7 is supported in full? I don't have an installation to test it, I can test no regression with 5.6

**Comment by SteveIreland** [comment]
I'm, of course, in no position to say that everything has been tested in full but I've been using MySQL 5.7 exclusively for the last couple of months and over the last couple of weeks I haven't stumbled across anything that isn't return a result and your last recommendations resulted in the results being valid.  So my vote is that the MySQL 5.7 issues are resolved ... but I'm continuing to test and will be setting up a production environment using MySQL 5.7 this weekend.

**Comment by odiea** [comment]
So far so good.
![2017-02-03 08_25_46-program manager](https://cloud.githubusercontent.com/assets/10962177/22596862/8322eac6-e9ea-11e6-9fd6-811283cf1fa6.png)

**Comment by daN4cat** [comment]
Good good

**Comment by waleogun68** [comment]
Hello, can you help with this 2 db Syntax in codeigniter?

I have tried  this "Where Not" here but not working
	public function checkReservation($newStart,$newEnd,$id,$newResource){
  	$stmt = $this->db->query('SELECT * FROM '. $this->db->dbprefix('reservations'). ' WHERE NOT ((end <= '.$newStart.') OR (start >= '.$newEnd.')) AND id <> '.$id.' AND room_id = '.$newResource);
               
	return $stmt->num_rows();
  	}

And also this not working too
 public function checkReservation($newStart,$newEnd,$id,$newResource)
{
        $stmt = $this->db->query('SELECT * FROM '. $this->db->dbprefix('reservations'));
        $this->db->group_start();
        $this->db->where('end >=', $shift_start);
       $this->db->where('start <=', $shift_end);
        $this->db->where('end >=', $newStart);
        $this->db->where('start <=', $newEnd);
        $this->db->group_end();
        $this->db->where('id <>',$id);
        $this->db->where('room_id',$newResource);
        return $stmt->num_rows();
  }

I tried also group start and end without any success.
   	public function getallReservations($start,$end)
	{
        $stmt=$this->db->query('SELECT * FROM '. $this->db->dbprefix('reservations'));
        $this->db->group_start();
        $this->db->where_not_in('end <=', $start);
        $this->db->or_where('start >=', $end);
        $this->db->group_end();
       
	$data = $stmt->result_array();
	return $data;
  	}

This Code is not also working
   	public function getallReservations($start,$end)
	{
  	$stmt = $this->db->query('SELECT * FROM '. $this->db->dbprefix('reservations'). ' WHERE NOT ((end >= '.$_POST['start'].') & (start >= '.$_POST['end'].'))'); 
	$data = $stmt->result_array();
	return $data;
  	}

**Comment by daN4cat** [comment]
Have you tried your SQL query using phpMyAdmin before moving to the code version?

**Comment by waleogun68** [comment]
Thanks for your response,

No, I have not but I have this php code that works without any problem,

public function getallReservations($start,$end){
  	$stmt = $this->db->query("SELECT * FROM reservations WHERE NOT ((end >= '".$_POST['start']."') & (start >= '".$_POST['end']."'))");
  	$data = $stmt->result_array();
  	return $data;
  }

and
 public function checkReservation($newStart,$newEnd,$id,$newResource){
  	$stmt = $this->db->query("SELECT * FROM reservations WHERE NOT ((end <= '".$newStart."') OR (start >= '".$newEnd."')) AND id <> ".$id." AND room_id = ".$newResource);
	return $stmt->num_rows();
  }

But I have issue when using Codeigniter dbprefix Syntax. 

Thanks Adewale

**Comment by daN4cat** [comment]
So it's a question to Codeigniter not for us.
